Convert the JSON code to use D slices instead of pointers

Geod24 opened the pull request:
Since we want to get rid of strlen, I tried to convert a simple global to a D slice and was blocked by the JSON module. So here is the module conversion to D.
Some simple things were added (e.g. `RootObject.toString` which just `strlen` on `toChars`), but I avoided changing the AST as much as possible, so things like `kind()` and `Loc.filename` can be done in another PR.

Geod24 committed:
ModuleDeclaration: Provide a toString method

Geod24 committed:
hdrgen: Add some `xxxToString` methods to wrap `xxxToChars`

strlen must die, right ?

Geod24 committed:
RootObject: Add a toString method

At the moment it is simply doing an `strlen` on the result of `toChars`,
but as more code uses it, we can start to move away from that.
Another reason to do it this way is because code can rely on the string
being null terminated, so this way `x.toString().ptr` is still a drop-in
replacement of `x.toChars()`.

Geod24 committed:
Remove trailing whitespace

Geod24 committed:
Convert the JSON code to use D slices instead of pointers

The conversion is pretty straightforward, but required in order to convert globals.
Since globals are still C-style string, a local toDString is used.
It can later be removed as globals are converted to D slices.

WalterBright commented on src/dmd/json.d:
```diff
@@ -391,26 +397,26 @@ public:
         if (type)
         {
-            property(name, type.toChars());
+            property(name, RootObject.toString(type));
```
Not too thrilled about this. It should be type.toString(). Yes, the C++ code cannot call it, but all you need to do is put a dummy function declaration there on the C++ side so the vtbl[] entries line up.

Geod24 replied:
Fixed

JinShil commented on src/dmd/dmodule.d:
```diff
@@ -1337,19 +1337,24 @@ struct ModuleDeclaration
         return buf.extractString();
     }
+
+    extern (D) const(char)[] toString() const
```
DDoc?

WalterBright replied:
As I mentioned before, toString() is used pervasively in dmd, and as long as it does what the other toString()'s do, adding a Ddoc for it is redundant.

rainers replied:
While I agree that a comment is not really necessary here, you have to add some ddoc comment to make the function appear in the documentation at all. Not sure if that is a good thing as you have to litter the code with empty comments this way. Is there an option to put all (public) symbols into the documentation?

Geod24 replied:
For the record I added a (trivial) DDOC comment just for this reason (documentation)

JinShil commented on src/dmd/hdrgen.d:
```diff
@@ -3347,6 +3347,12 @@ extern (C++) const(char)* stcToChars(ref StorageClass stc)
     return null;
 }
 
+extern (D) const(char)[] stcToString(ref StorageClass stc)
```
DDoc?

JinShil commented on src/dmd/hdrgen.d:
```diff
@@ -3355,6 +3361,12 @@ extern (C++) void trustToBuffer(OutBuffer* buf, TRUST trust)
+    return trustToString(trust).ptr;
+}
+
+extern (D) string trustToString(TRUST trust)
```
DDoc?

JinShil commented on src/dmd/hdrgen.d:
```diff
@@ -3417,6 +3429,12 @@ extern (C++) void protectionToBuffer(OutBuffer* buf, Prot prot)
+    return protectionToString(kind).ptr;
+}
+
+extern (D) string protectionToString(Prot.Kind kind)
```
DDoc?

jacob-carlborg commented on src/dmd/json.d:
```diff
@@ -361,22 +368,21 @@ public:
+            if (auto filename = toDString(loc.filename))
             {
-                if (!this.filename || strcmp(filename, this.filename))
+                if (!this.filename.length || filename != this.filename)
```
I'm guessing you're trying keep the new to code be as close as possible to the old code but this first expression (`!this.filename.length`) isn't needed anymore?

Geod24 commented:
Updated according to feedback

ibuclaw commented on src/dmd/dmodule.d:
```diff
@@ -1337,19 +1337,25 @@ struct ModuleDeclaration
         return buf.extractString();
     }
+
+    /// Provide an human readable representation
```
Provide a human

ibuclaw commented on src/dmd/hdrgen.d:
```diff
@@ -3347,14 +3347,37 @@ extern (C++) const(char)* stcToChars(ref StorageClass stc)
+
+/**
+ * Returns:
+ *   an human readable representation of `stc`
```
A human

ibuclaw commented on src/dmd/hdrgen.d:
```diff
@@ -3347,14 +3347,37 @@ extern (C++) const(char)* stcToChars(ref StorageClass stc)
 
+/**
+ * Returns:
+ *   an human readable representation of `trust`,
```
A human

ibuclaw commented on src/dmd/hdrgen.d:
```diff
@@ -3416,7 +3439,18 @@ extern (C++) void protectionToBuffer(OutBuffer* buf, Prot prot)
 
+/**
+ * Returns:
+ *   an human readable representation of `kind`
```
A human

ibuclaw commented on src/dmd/json.d:
```diff
@@ -83,11 +91,10 @@ public:
     }
 
-    void stringPart(const(char)* s)
+    extern(D) void stringPart(const(char)[] s)
```
Wouldn't all non-visit members of this visitor be private? Ergo you can just add `extern(D):` at the top of this body rather than adding it for each individual function.

Think about it for future refactor.

Geod24 replied:
True that. At the moment I'm putting `extern(D)` right and left (got more slice-ification coming), but I also had weird bugs (compiler SEGV) being triggered in the past some I decided to go for verbose and simple. But when enough is converted, we surely can do that.

ibuclaw commented on src/dmd/root/outbuffer.d:
```diff
@@ -459,4 +459,3 @@ char[] unsignedToTempString(ulong value, char[] buf, uint radix = 10) @safe
     } while (value);
     return buf[i .. $];
 }
-
```
Seems like a leftover from a previous attempt here.

Geod24 replied:
It is, but also its own commit

Geod24 commented:
Fixed typos

wilzbach commented on src/dmd/json.d:
```diff
@@ -47,13 +48,20 @@ private extern (C++) final class ToJsonVisitor : Visitor
+    static extern(D) const(char)[] toDString (const(char)* s)
+    {
+        return s ? s[0 .. strlen(s)] : null;
+    }
```
Why not put this as a global function into `utils` or similar? It's used at other places too and doesn't really fit in `ToJsonVisitor`

Geod24 replied:
Didn't expect it to be so widely used :(
Also I'm wary that `dmd.utils` will be imported everywhere soon.
But let's give it a shot...

wilzbach commented on src/dmd/hdrgen.d:
```diff
@@ -3347,14 +3347,37 @@ extern (C++) const(char)* stcToChars(ref StorageClass stc)
+extern (D) const(char)[] stcToString(ref StorageClass stc)
+{
+    auto p = stcToChars(stc);
+    return p[0 .. strlen(p)];
```
Could use `toDString` from below.

wilzbach commented on src/dmd/dmodule.d:
```diff
@@ -1337,19 +1337,25 @@ struct ModuleDeclaration
+    extern (D) const(char)[] toString() const
+    {
+        auto p = this.toChars();
+        return p[0 .. strlen(p)];
```
Could use `toDString` from below.

wilzbach commented on src/dmd/json.d:
```diff
@@ -526,20 +532,20 @@ public:
-        property("kind", s.kind());
-        property("comment", s.comment);
+        property("kind", toDString(s.kind()));
+        property("comment", toDString(s.comment));
```
Nit: another advantage of making `toDString` a global function is that you can do the more natural: `s.comment.toDString()`

wilzbach commented on src/dmd/root/rootobject.d:
```diff
@@ -62,6 +62,14 @@ extern (C++) class RootObject
+    {
+        import core.stdc.string : strlen;
+        auto p = this.toChars();
+        return p[0 .. strlen(p)];
```
Could use `toDString` from above.

Geod24 replied:
Here I choose not to, as we want `rootobject` to depends on as little as possible.

Geod24 committed:
dmd.utils: Add toDString to slice a C-string

Geod24 commented:
Addressed @wilzbach 's comments

JinShil reviewed (approved):
Still looking good.

Geod24 committed:
Move DArray definition to DCompat

If we want to use DArray from dmd.root it cannot reside in globals.h,
so it is moved to a standalone header.

Geod24 commented:
Since this has gotten a bit more reviews, and one might be wondering where I am going, I added a bit more diff than originally intended:
- Commit 0c91769 moves `DArray` from `globals.h` to a standalone header, because we don't want `object.h` importing `globals.h`
- I reworked `RootObject.toString` to be callable from C++ and added a test.

As for the end goal is, [it's to be able to do this](https://github.com/Geod24/dmd/commit/1a740a2e98294e6d5430659d4ca358d5f05bbdda#diff-3ea076fe941194ddccb18fc686689e4eL148) gradually.

jacob-carlborg commented on src/tests/cxxfrontend.c:
```diff
@@ -212,6 +212,11 @@ void test_semantic()
     semantic3(m, NULL);
 
+    // Basic test for D slice passing
+    DArray<const char> sl = m->toString();
```
Does this work on Windows? I’ve been told D arrays are returned in registers while structs of this size are not on Windows.

Geod24 replied:
I should've known it was too good to be true.
I'll wait for the auto-tester to tell me what's the situation with Windows, but it fails on 32 bits.
So I guess I will take that test out for now :'(

Geod24 replied:
Hum, the C++ frontend tests only run on POSIX anyway...

jacob-carlborg replied:
For reference: https://github.com/dlang/dmd/pull/8120#discussion_r200835776.

jacob-carlborg commented on src/dmd/utils.d:
```diff
@@ -132,3 +132,9 @@ extern (C++) void escapePath(OutBuffer* buf, const(char)* fname)
 }
+
+/// Slices a `\0`-terminated C-string, excluding the terminator
+const(char)[] toDString (const(char)* s) pure nothrow @nogc
```
Why not `toString`?

Geod24 replied:
`toString` is to provide a string representation, this is to convert a `const(char)*` to a `const(char)[]`.

Geod24 commented:
@wilzbach : 
```
dmd/root/rootobject.d(66:29)[warn]: Methods 'opCmp', 'toHash', 'opEquals', 'opCast', and/or 'toString' are non-const.
```

What's the deal with that warning ?? `extern(C++)` classes don't inherit from `Object` so it doesn't make sense to require them to be `const`. And I currently cannot make it `const`.

wilzbach commented:
> What's the deal with that warning ?? extern(C++) classes don't inherit from Object so it doesn't make sense to require them to be const. 

Yeah that's the newly added Dscanner static code analysis check and the idea is that the checks slowly allow to modernize DMD / enforce some automatic code checking.
I guess the assumption behind this check was that "opEquals, opCmp, toHash, and toString are either const, immutable or inout" because they shouldn't be changing any data.

> And I currently cannot make it const.

tl;dr: in this case, add your module to this list:

https://github.com/dlang/dmd/blob/48726c875fbfe16fd93051570d149e893d2dc3dc/src/.dscanner.ini#L118

jacob-carlborg commented on src/dmd/root/object.h:
```diff
@@ -57,6 +58,9 @@ class RootObject
     virtual const char *toChars();
+    /// This function is `extern(D)` and should not be called from C++,
+    /// as the ABI does not match on some platforms
+    virtual DArray<const char> toString();
```
Is it possible to implement this on the C++ and have the implementation assert?

Geod24 replied:
I don't think so.

Geod24 commented:
@wilzbach : Thanks, glad to see there's a way out of it!

wilzbach commented:
@Geod24 I don't know exactly what's up with SemaphoreCI, but https://github.com/dlang/dmd/pull/8560 (once merged back to master) will hopefully help a bit.

Geod24 commented:
Rebased on master, let's see.

JinShil reviewed (approved):
Yup, I still like it.

wilzbach reviewed (approved):
Yeah same here. One small step forward (:

Geod24 commented:
Anything else ? Got a few changes depending on this PR (to be precise, on `toDString`).

wilzbach commented:
I think this is just pending/blocked on Ian's change request. @ibuclaw are you okay with moving forward here?
